### PR TITLE
Siw authenticator txt 459619

### PR DESCRIPTION
**Pull request by lokeshhemrajani-okta**
## Description:
- Updated the authenticator text for forgot-password flow when challenge is sent
- Hide the 'Click the link..' text when there is no email magic link in IdxResponse->currentAuthenticatorEnrollment
- Backend story for this OKTA-454634

## PR Checklist

- [ ] Have you verified the basic functionality for this change?

- [ ] Added unit tests?

- [ ] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Reviewers:

@okta/ciamx 

### Issue:

- [OKTA-459619](https://oktainc.atlassian.net/browse/OKTA-459619)

https://user-images.githubusercontent.com/92340204/152280100-7a57327f-f898-4ca0-a49f-16ffb0525b46.mov

### CODE REVIEW

File: packages/@okta/i18n/src/properties/login.properties [hunk 1]
```diff
@@ -929,10 +929,11 @@ oie.email.verify.primaryButton = Send me an email
 oie.email.verify.subtitleWithEmailAddress = Verify with an email link or enter a code sent to <$1>{0}</$1>.
 oie.email.verify.subtitleWithoutEmailAddress = Verify with an email link or enter a code sent to your email.
 # {0} is a user's email address
-oie.email.verify.alternate.magicLinkToEmailAddress = An email magic link was sent to <$1>{0}</$1>. 
+oie.email.verify.alternate.magicLinkToEmailAddress = We sent an email to <$1>{0}</$1>. 
 oie.email.verify.alternate.magicLinkToYourEmail = An email magic link was sent to your email. 
-oie.email.verify.alternate.instructions = Click the link in the email or enter the code below to continue.
+oie.email.verify.alternate.instructions = Click the verification link in your email to continue or enter the code below.
 oie.email.verify.alternate.showCodeTextField = Enter a code from the email instead
+oie.email.verify.alternate.verificationCode.instructions = Enter the verification code in the text box.
```
Comment (justinabrokwah-okta) [inline packages/@okta/i18n/src/properties/login.properties:936]: @IanHakes-okta ⬆️ some user facing strings here

**Review by justinabrokwah-okta — COMMENTED** [review]
Changes look good overall, can you add a screenshot or video for us to see?

**Comment by lokeshhemrajani-okta** [comment]
@justinabrokwah-okta Will add video soon

File: packages/@okta/i18n/src/properties/login.properties [hunk 2]
```diff
@@ -929,10 +929,11 @@ oie.email.verify.primaryButton = Send me an email
 oie.email.verify.subtitleWithEmailAddress = Verify with an email link or enter a code sent to <$1>{0}</$1>.
 oie.email.verify.subtitleWithoutEmailAddress = Verify with an email link or enter a code sent to your email.
 # {0} is a user's email address
-oie.email.verify.alternate.magicLinkToEmailAddress = An email magic link was sent to <$1>{0}</$1>. 
+oie.email.verify.alternate.magicLinkToEmailAddress = We sent an email to <$1>{0}</$1>.
```
Comment (zeusandrade-okta) [inline packages/@okta/i18n/src/properties/login.properties:932]: Why are we changing the existing UI strings?

Comment (pradeepdewda-okta) [reply packages/@okta/i18n/src/properties/login.properties:932]: @lokeshhemrajani-okta are we not going to show `An email magic link was sent` in any other case now? are we scrapping it ?

Comment (lokeshhemrajani-okta) [reply packages/@okta/i18n/src/properties/login.properties:932]: @IanHakes-okta can you please reply on this? I updated as per comment in the story OKTA-454634

Comment (IanHakes-okta) [reply packages/@okta/i18n/src/properties/login.properties:932]: Change based on 454634 - @bhagyaprabhakar-okta asked for text to resolve the scenario where "the customer is able to customize the email template and not have the email link at all.  In this case, the authenticator screen should not include any instructions to click link in email."

File: src/v2/view-builder/views/email/ChallengeAuthenticatorEmailView.js [hunk 3]
```diff
@@ -61,15 +68,24 @@ const Body = BaseAuthenticatorEmailForm.extend(
...
-      this.add(EnterCodeLink, {
-        prepend: true,
-        selector: '.o-form-error-container',
-      });
+      const useEmailMagicLink = this.options.appState.get('currentAuthenticatorEnrollment')?.
+      contextualData?.useEmailMagicLink;
+      const useEmailMagicLinkValue = useEmailMagicLink !== undefined ? useEmailMagicLink : true;
+
+      //If email magic link exist then show button with text, else toggle button to show text box
+      if (useEmailMagicLinkValue) {
```
Comment (xifengfang-okta) [inline src/v2/view-builder/views/email/ChallengeAuthenticatorEmailView.js:76]: I think you can just use if (!!useEmailMagicLink) and line 73 is not necessary,

Comment (lokeshhemrajani-okta) [reply src/v2/view-builder/views/email/ChallengeAuthenticatorEmailView.js:76]: !! -> does this check for undefined and true both?

Comment (xifengfang-okta) [reply src/v2/view-builder/views/email/ChallengeAuthenticatorEmailView.js:76]: I think so,
` value     │  !value  │  !!value
━━━━━━━━━━━┿━━━━━━━━━━┿━━━━━━━━━━━
 false     │ ✔ true   │   false
 true      │   false  │ ✔ true
 null      │ ✔ true   │   false
 undefined │ ✔ true   │   false
 0         │ ✔ true   │   false
 -0        │ ✔ true   │   false
 1         │   false  │ ✔ true
 -5        │   false  │ ✔ true
 NaN       │ ✔ true   │   false
 ''        │ ✔ true   │   false
 'hello'   │   false  │ ✔ true`

Comment (xifengfang-okta) [reply src/v2/view-builder/views/email/ChallengeAuthenticatorEmailView.js:76]: I think so
Screen Shot 2022-02-01 at 5 18 28 PM

Comment (lokeshhemrajani-okta) [reply src/v2/view-builder/views/email/ChallengeAuthenticatorEmailView.js:76]: this wont work for me because if you see the logic I wrote, when its undefined I want it as true. But as per chart !! returns false.

File: packages/@okta/i18n/src/properties/login.properties (hunk 2 above)
Comment (pradeepdewda-okta) [reply packages/@okta/i18n/src/properties/login.properties:932]: @IanHakes-okta @bhagyaprabhakar-okta question is are we scrapping `An email magic link was sent` text from all use cases ?

@lokeshhemrajani-okta must verify this in okta-core before merging to widget.. there might selenium test failures for this text change.

File: playground/mocks/data/idp/idx/authenticator-verification-email-without-emailmagiclink.json [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+          {
+            "name":"stateHandle",
+            "required":true,
+            "value":"02WTSGqlHUPjoYvorz8T48txBIPe3VUisrQOY4g5N8",
+            "visible":false,
+            "mutable":false
+          }
+        ]
+      },
+      "contextualData": {
```
Comment (pradeepdewda-okta) [inline playground/mocks/data/idp/idx/authenticator-verification-email-without-emailmagiclink.json:110]: just for my information .. how did we finally decide to add this here ?

File: packages/@okta/i18n/src/properties/login.properties (hunk 2 above)
Comment (bhagyaprabhakar-okta) [reply packages/@okta/i18n/src/properties/login.properties:932]: Yes.  The intention is to use "active voice" and have a more direct instruction to the user.

File: src/v2/view-builder/views/email/ChallengeAuthenticatorEmailView.js [hunk 5]
```diff
@@ -86,8 +88,16 @@ const Body = BaseAuthenticatorEmailForm.extend(
 
     postRender() {
       BaseAuthenticatorEmailForm.prototype.postRender.apply(this, arguments);
-
-      this.showCodeEntryField(false);
+      const useEmailMagicLink = this.options.appState.get('currentAuthenticatorEnrollment')?.
```
Comment (pradeepdewda-okta) [inline src/v2/view-builder/views/email/ChallengeAuthenticatorEmailView.js:91]: create a new local function for `useEmailMagicLink()` that returns true and false.. to reuse it.

File: src/v2/view-builder/views/email/ChallengeAuthenticatorEmailView.js [hunk 6]
```diff
@@ -61,22 +68,36 @@ const Body = BaseAuthenticatorEmailForm.extend(
       const { email } =
         this.options.currentViewState.relatesTo?.value?.profile || {};
 
-      this.add(EnterCodeLink, {
-        prepend: true,
-        selector: '.o-form-error-container',
-      });
+      const useEmailMagicLink = this.options.appState.get('currentAuthenticatorEnrollment')?.
+      contextualData?.useEmailMagicLink;
```
Comment (pradeepdewda-okta) [inline src/v2/view-builder/views/email/ChallengeAuthenticatorEmailView.js:72]: to get `contextualData` add a function like this
```
  getContextualData() {
    return this.options.appState.get('currentAuthenticator').contextualData;
  },
 ```

**Comment by lokeshhemrajani-okta** [comment]
@pradeepdewda-okta refactored

**Comment by lokeshhemrajani-okta** [comment]
@justinabrokwah-okta Attached is the recording.

File: playground/mocks/config/responseConfig.js [hunk 7]
```diff
@@ -115,6 +115,7 @@ const idx = {
   '/idp/idx/identify': [
     'authenticator-verification-email',
     // 'authenticator-enroll-select-authenticator',
+    // 'authenticator-verification-email-without-emailmagiclink',
```
Comment (justinabrokwah-okta) [inline playground/mocks/config/responseConfig.js:118]: Just curious, what is this line commented out for?

Comment (lokeshhemrajani-okta) [reply playground/mocks/config/responseConfig.js:118]: I added new json and added here if we want to test without emailmagiclink scenario so that everyone can see and use it rather than creating/searching for one.

File: package.json [hunk 8]
```diff
@@ -2,7 +2,7 @@
   "private": true,
   "name": "@okta/okta-signin-widget",
   "description": "The Okta Sign-In Widget",
-  "version": "6.0.0",
+  "version": "6.1.0",
```
Comment (pradeepdewda-okta) [inline package.json:5]: you i don't think you need to make this change..

cc: @aarongranick-okta

**Comment by pradeepdewda-okta** [comment]
@lokeshhemrajani-okta please rebase correctly so that only your change reflect in the PR.

File: package.json (hunk 8 above)
Comment (aarongranick-okta) [inline package.json:5]: please revert this change

**Review by aarongranick-okta — CHANGES_REQUESTED** [review]
please remove the version bump. this is managed by widget coodinator

**Comment by lokeshhemrajani-okta** [comment]
@aarongranick-okta @pradeepdewda-okta Change of main branch caused issue. I have resolved it.

**Review by justinabrokwah-okta — APPROVED** [review]
LGTM 🚀

**Review by pradeepdewda-okta — CHANGES_REQUESTED** [review]
Add a testcafe test that test for `click to link` is not shown up when use useEmailMagicLinkValue exists..

File: src/v2/view-builder/views/email/ChallengeAuthenticatorEmailView.js [hunk 9]
```diff
@@ -24,15 +24,22 @@ const CheckYourEmailTitle = View.extend({
...
-    }}
+    
+    {{#if useEmailMagicLinkValue}}
+      {{i18n 
+        code="oie.email.verify.alternate.instructions" 
+        bundle="login" 
+      }}
+    {{else}}
+      {{i18n 
+        code="oie.email.verify.alternate.verificationCode.instructions"
```
[the remark this answers is not in the document]
Comment (lokeshhemrajani-okta) [reply src/v2/view-builder/views/email/ChallengeAuthenticatorEmailView.js:35]: Done

**Comment by pradeepdewda-okta** [comment]
@lokeshhemrajani-okta do we have green bacon in okta-core for string changes ?

**Comment by lokeshhemrajani-okta** [comment]
@pradeepdewda-okta we do. https://bacon-go.aue1e.saasure.net/commits?artifact=monolith&branch=lh-fgtpwd-text-454634&page=1&pageSize=6

**Review by pradeepdewda-okta — APPROVED** [review]
